src/Uno.UI.Runtime.Skia.Wpf/WPF/WpfDragDropExtension.cs

```diff
@@ -0,0 +1,279 @@
+			var data = ToDataPackage(e.Data);
+			var allowedOperations = ToDataPackageOperation(e.AllowedEffects);
+			var info = new CoreDragInfo(src, data.GetView(), allowedOperations);
+
+			CoreDragDropManager.GetForCurrentView()?.DragStarted(info);
+
+			// Note: No needs to _manager.ProcessMove, the DragStarted will actually have the same effect
+		}
+
+		private void OnHostDragOver(object sender, DragEventArgs e)
+			=> e.Effects = ToDropEffects(_manager.ProcessMoved(new DragEventSource(_fakePointerId, e)));
+
+		private void OnHostDragLeave(object sender, DragEventArgs e)
+			=> e.Effects = ToDropEffects(_manager.ProcessAborted(new DragEventSource(_fakePointerId, e)));
+
+		private void OnHostDrop(object sender, DragEventArgs e)
+			=> e.Effects = ToDropEffects(_manager.ProcessDropped(new DragEventSource(_fakePointerId, e)));
+
+		public void StartNativeDrag(CoreDragInfo info)
+			=> WpfHost.Current.Dispatcher.InvokeAsync(async () =>
+			{
+				try
+				{
+					var data = await ToDataObject(info.Data, CancellationToken.None);
+					var effects = ToDropEffects(info.AllowedOperations);
+
+					DragDrop.DoDragDrop(WpfHost.Current, data, effects);
+				}
+				catch (Exception e)
+				{
+					this.Log().Error("Failed to start native Drag and Drop.", e);
+				}
+			});
+
+		private static DataPackageOperation ToDataPackageOperation(DragDropEffects wpfOp)
+			=> (DataPackageOperation)((int)wpfOp) & (DataPackageOperation.Copy | DataPackageOperation.Move | DataPackageOperation.Link);
+
+		private static DragDropEffects ToDropEffects(DataPackageOperation uwpOp)
+			=> (DragDropEffects)((int)uwpOp) & (DragDropEffects.Copy | DragDropEffects.Move | DragDropEffects.Link);
+
+		private static DataPackage ToDataPackage(IDataObject src)
+		{
+			var dst = new DataPackage();
+			var text = src.GetData(DataFormats.Text) as string
+				?? src.GetData(DataFormats.UnicodeText) as string
+				?? src.GetData(DataFormats.OemText) as string
+				?? src.GetData(DataFormats.StringFormat) as string;
+
+			if (text.HasValueTrimmed())
+			{
+				if (Uri.IsWellFormedUriString(text, UriKind.Absolute))
+				{
+					dst.SetWebLink(new Uri(text!));
+				}
+				else
+				{
+					dst.SetText(text!);
+				}
+			}
+
+			if (src.GetData(DataFormats.Html) is string html)
+			{
+				dst.SetHtmlFormat(html);
+			}
+
+			if (src.GetData(DataFormats.Rtf) is string rtf)
+			{
+				dst.SetRtf(rtf);
+			}
+
+			if (src.GetData(DataFormats.Bitmap) is BitmapSource bitmap)
+			{
+				dst.SetBitmap(new RandomAccessStreamReference(async ct =>
+				{
+					var copy = new MemoryStream();
+					var encoder = new BmpBitmapEncoder();
+					encoder.Frames.Add(BitmapFrame.Create(bitmap));
+					encoder.Save(copy);
+					copy.Position = 0;
+
+					return copy.AsRandomAccessStream().TrySetContentType("image/bmp");
+				}));
+			}
+
+			if (src.GetData(DataFormats.FileDrop) is string[] files)
+			{
+				dst.SetStorageItems(files.Select(StorageFile.GetFileFromPath));
```

Comment:
Doing this when the DataPackage is created means that the file will be read as soon as the cursor enters the Uno app. This **may** not be performance friendly (although the file isn't copied here) and it would be better to use `DataPackage.SetDataProvider()`. For the macOS platform I use the following code:

```csharp
dataPackage.SetDataProvider(
    StandardDataFormats.StorageItems,
    async cancellationToken =>
    {
        // Convert rom a temp Url (see above example) into an absolute file path
        var fileUrl = new NSUrl(tempFileUrl);
        var file = await StorageFile.GetFileFromPathAsync(fileUrl.FilePathUrl.AbsoluteString);

        var storageItems = new List<IStorageItem>();
        storageItems.Add(file);

        return storageItems.AsReadOnly();
        });
```

Reply:
With the [latest version of the `StorageFile`](https://github.com/unoplatform/uno/blob/master/src/Uno.UWP/Storage/StorageFile.cs#L24), this will actually only create some instances of `StorageFile` (and it does not create a `FileInfo` which would indeed have touched the file system). It's even using an internal sync overload which makes it the most lightweight possible I think.

Did you notice any perf issue with this one or it's just theoretical?

Reply:
This one was just theoretical :) I did it differently on macOS just to be on the extra-safe side. Perhaps that wasn't necessary but I didn't know the internals of how StorageFile works on all platforms and could see it accessing the file system so wanted to be careful.